File: week_5/BOJ_1654.py
import sys
input=sys.stdin.readline
k, n=map(int, input().split())
lines=[list(int(input()) for _ in range(n))]

start, end=1, max(lines)
while start<=end:
    mid=(start+end)//2
    cnt=0
    for line in lines :
        cnt+=(line//mid)
    if cnt>=n:
        start=mid+1
    else:
        end=mid-1
print(end)

File: week_05/minji/BOJ_1654.java
import java.util.Scanner;

public class BOJ_1654 {
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int k=sc.nextInt();
        int n=sc.nextInt();
        int[] lines=new int[n];
        int end=0;
        for(int i=0; i<n; i++){
            lines[i]=sc.nextInt();
            if(end<lines[i]) end=lines[i];
        }
        int start=0;
        int mid=0;
        while(start<=end){
            mid=(start+end)/2;
            int count=0;
            for(int line : lines){
                count+=line/mid;
            }
            if(count>=n){
                start=mid+1;
            }else{
                end=mid-1;
            }
        }
        System.out.println(end);
    }
}
